test/partition.test.js

```diff
@@ -28,141 +29,188 @@ describe('cloudant - partitioned db', () => {
+      });
+      db.automigrate('Product', (err) => {
+        if (err) return done(err);
+        connector.getIndexes(connector.getDbName(connector), (e, results) => {
+          if (e) return done(e);
+          const indexes = results.indexes;
+          const indexName = 'prodName3_index';
+          should.exist(indexes);
+
+          const index = _.find(indexes, function(index) {
+            return index.name === indexName;
+          });
+          should.exist(index);
+          index.name.should.equal(indexName);
+          index.def.fields[0]['prodName'].should.equal('asc');
+          index.def.fields[1][DEFAULT_MODEL_VIEW].should.equal('asc');
+          // should be a global index
+          index.partitioned.should.equal(false);
+          done();
+        });
+      });
     });
-    db.automigrate('Product', (err) => {
-      if (err) return done(err);
-      connector.getIndexes(connector.getDbName(connector), (e, results) => {
-        if (e) return done(e);
-        const indexes = results.indexes;
-        const indexName = 'prodName3_index';
-        should.exist(indexes);
+  });
 
-        const index = _.find(indexes, function(index) {
-          return index.name === indexName;
+  context('findAll tests', ()=> {
+    it('find all records by partition key from option', (done) => {
+      Product = db.define('Product', {
+        name: {type: String},
+        tag: {type: String},
+      }, {
+        forceId: false,
+        indexes: {
+          'product_name_index': {
+            partitioned: true,
+            keys: {
+              city: 1,
+            },
+          },
+        },
+      });
+      db.automigrate('Product', () => {
+        Product.create(SEED_DATA, function(err) {
+          if (err) return done(err);
+          Product.find({where: {tag: 'food'}}, {partitionKey: 'toronto'},
+            (err, results) => {
+              if (err) return done(err);
+              should.exist(results);
+              results.length.should.equal(2);
+              const resultTaggedFood = _.filter(results,
+                function(r) {
+                  return r.tag === 'food';
+                });
+              resultTaggedFood.length.should.equal(2);
+              done();
+            });
         });
-        should.exist(index);
-        index.name.should.equal(indexName);
-        index.def.fields[0]['prodName'].should.equal('asc');
-        index.def.fields[1][DEFAULT_MODEL_VIEW].should.equal('asc');
-        // should be a global index
-        index.partitioned.should.equal(false);
-        done();
       });
     });
   });
 });
+
+const SEED_DATA = [
+  {id: `toronto: ${uuid()}`, name: 'beer', tag: 'drink'},
+  {id: `toronto: ${uuid()}`, name: 'salad', tag: 'food'},
+  {id: `toronto: ${uuid()}`, name: 'soup', tag: 'food'},
+  {id: `london: ${uuid()}`, name: 'beer', tag: 'drink'},
+  {id: `london: ${uuid()}`, name: 'salad', tag: 'food'},
+  {id: `london: ${uuid()}`, name: 'salad', tag: 'food'},
+];
```

Comment:
What if I have a 7th record `{id: toronto: ${uuid()}, name: 'dessert', tag: 'food'},`, would it be included to the result of the above test case?

Reply:
yep, because its partitionKey is `toronto`(matches your partition key) and is tagged as `food`(match your query).

Reply:
I see, the name `partitionKey` always makes me think of operations based on index.

Reply:
Yep they are related :) 
> The benefit of partitioned databases is that secondary indices can be significantly more efficient when locating matching documents since their entries are contained within their partition. This means a given secondary index read will only scan a single partition range instead of having to read from a copy of every shard.

My understanding is index is a secondary optimization for search.